### ISSUE TITLE
"Schema must be union" error with Native Avro File Reader

### ISSUE THREAD

**Issue by kekwan** [trinodb/trino#20310]
We recently upgraded from Trino 405 to 427. https://github.com/trinodb/trino/pull/17221 makes our Hive external table incompatible with the latest version.

Our Hive external table is defined as follows. 

```
CREATE EXTERNAL TABLE `avro_logs`(
  `value` binary COMMENT '',
  `timestamp` timestamp COMMENT '')
PARTITIONED BY (
  `etl_ts_` timestamp,
  `symlink_try_number` int,
  `batch_number` int)
ROW FORMAT SERDE
  'org.apache.hadoop.hive.serde2.avro.AvroSerDe'
STORED AS INPUTFORMAT
  'org.apache.hadoop.hive.ql.io.SymlinkTextInputFormat'
OUTPUTFORMAT
  'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat'
LOCATION
  's3://bucket'
TBLPROPERTIES (
  'avro.schema.literal'='{\n   "type":"record",\n   "name":"KafkaRecord",\n   "namespace":"com.test",\n   "fields":[\n      {\n         "name":"value",\n         "type":"bytes"\n      },\n      {\n         "name":"timestamp",\n         "type":[\n            "null",\n            {\n               "type":"long",\n               "connect.version":1,\n               "connect.name":"org.apache.kafka.connect.data.Timestamp",\n               "logicalType":"timestamp-millis"\n            }\n         ],\n         "default":null\n      }\n   ],\n   "connect.version":1,\n   "connect.name":"com.test.KafkaRecord"\n}',
  'presto_query_id'='20231215_211132_27985_4ahdm',
  'presto_version'='405-sfdc.40-sfdc.40-332b5f0-dirty',
  'transient_lastDdlTime'='1702674692')
```
We are now getting this error
```
com.google.common.base.VerifyException: Schema must be union
	at com.google.common.base.Verify.verify(Verify.java:126)
	at io.trino.hive.formats.avro.AvroTypeUtils.isSimpleNullableUnion(AvroTypeUtils.java:91)
	at io.trino.hive.formats.avro.AvroPageDataReader.createBlockBuildingDecoderForAction(AvroPageDataReader.java:196)
	at io.trino.hive.formats.avro.AvroPageDataReader$RowBlockBuildingDecoder.<init>(AvroPageDataReader.java:553)
	at io.trino.hive.formats.avro.AvroPageDataReader$RowBlockBuildingDecoder.<init>(AvroPageDataReader.java:529)
	at io.trino.hive.formats.avro.AvroPageDataReader.initialize(AvroPageDataReader.java:97)
	at io.trino.hive.formats.avro.AvroPageDataReader.setSchema(AvroPageDataReader.java:107)
	at org.apache.avro.file.DataFileStream.initialize(DataFileStream.java:145)
	at org.apache.avro.file.DataFileReader.<init>(DataFileReader.java:143)
	at org.apache.avro.file.DataFileReader.<init>(DataFileReader.java:130)
	at io.trino.hive.formats.avro.AvroFileReader.<init>(AvroFileReader.java:74)
	at io.trino.plugin.hive.avro.AvroPageSource.<init>(AvroPageSource.java:50)
	at io.trino.plugin.hive.avro.AvroPageSourceFactory.createPageSource(AvroPageSourceFactory.java:175)
	at io.trino.plugin.hive.HivePageSourceProvider.createHivePageSource(HivePageSourceProvider.java:202)
	at io.trino.plugin.hive.HivePageSourceProvider.createPageSource(HivePageSourceProvider.java:137)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:48)
```

Is this correct behaviour? Why are schemas enforced to be Union type?

**Comment by kekwan** [comment]
cc @jklamer

**Comment by jklamer** [comment]
@kekwan looking into this, this does look like an incorrect check, but it does also point to a potentially breaking data issue that only recently became possible to work around.
Explanation:
TLDR: Might be able to work around by changing the `value` field in your  table schema to  `["null", "bytes"]`. This likely will not have an effect on your queries of this table.

Details:
This error happens when the schema that wrote the data does not match the schema reading the data. For this error, the schema that wrote the data is a union, and the schema reading is not. Educated guess is that the file has `value` field with type `["null", "bytes"]`. Reading `["null", "bytes"]` with `bytes` only recently became something we could support, by delaying errors until read time. If there are only bytes values in the written file, the reading should complete. Before, we threw an error because you can't read all values of type `["null", "bytes"]` with type `"bytes"`.
However, this check is also incorrect because there are situations in which coercion could happen for all values of the unions. For example: you can read  all values of `["string", "bytes"]` with `"bytes"`.

I'll have a PR out shortly to fix this issue. Try the workaround if possible to unblock.

**Comment by kekwan** [comment]
Thanks @jklamer for quickly looking into this issue and getting a fix! 

I can confirm your workaround. We have lots of users with this pattern so the fix will be useful for us. It may be difficult for us to update all our table schemas. 

Our table schema is defined like this:
```
{
    "type": "record",
    "name": "KafkaRecord",
    "namespace": "com.test",
    "fields": [
      {
        "name": "value",
        "type": "bytes"
      },
      {
        "name": "timestamp",
        "type": [
          "null",
          {
            "type": "long",
            "connect.version": 1,
            "connect.name": "org.apache.kafka.connect.data.Timestamp",
            "logicalType": "timestamp-millis"
          }
        ],
        "default": null
      }
    ],
    "connect.version": 1,
    "connect.name": "com.test.KafkaRecord"
  }
```
whereas the actual Avro schema from the data file is 
```
  {
    "type": "record",
    "name": "KafkaRecord",
    "namespace": "com.test",
    "fields": [
      {
        "name": "value",
        "type": [
          "null",
          "bytes"
        ],
        "default": null
      },
      {
        "name": "timestamp",
        "type": [
          "null",
          {
            "type": "long",
            "connect.version": 1,
            "connect.name": "org.apache.kafka.connect.data.Timestamp",
            "logicalType": "timestamp-millis"
          }
        ],
        "default": null
      }
    ],
    "connect.version": 1,
    "connect.name": "com.salesforce.KafkaRecord"
  }
```

**Comment by jklamer** [comment]
The change in the PR combined with this recent change should solve your issue: https://github.com/trinodb/trino/pull/20233

**Comment by jklamer** [comment]
Pr merged @kekwan. Next release should contain the fix. let me know if you have any more issues with the reader!

**Comment by kekwan** [comment]
thank you for the help! 

will cherry-pick these fixes internally and test it out!
